core/src/avm2/regexp.rs

```diff
@@ -204,17 +209,68 @@ impl<'gc> RegExp<'gc> {
         ret
     }
 
-    pub fn replace(
+    // Implements string.replace(regex, replacement) where the replacement is
+    // a function.
+    pub fn replace_fn(
+        &mut self,
+        activation: &mut Activation<'_, 'gc, '_>,
+        text: AvmString<'gc>,
+        f: &FunctionObject<'gc>,
+    ) -> AvmString<'gc> {
+        self.replace_with_fn(activation, &text, |activation, txt, m| {
+            let mut args = std::iter::once(Some(&m.range))
+                .chain((m.captures.iter()).map(|x| x.as_ref()))
+                .map(|o| match o {
+                    Some(r) => Value::from(AvmString::new(
+                        activation.context.gc_context,
+                        &txt[r.start..r.end],
+                    )),
+                    None => Value::from(AvmString::default()),
+                })
+                .collect::<Vec<_>>();
+            args.push(Value::from(m.range.start));
+            args.push(Value::from(*txt));
+
+            let r = f.call(activation.global_scope(), &args, activation);
+            return WString::from(
+                r.unwrap_or_else(|_| Value::String("".into()))
+                    .coerce_to_string(activation)
+                    .unwrap_or_default()
+                    .as_wstr(),
+            );
+        })
+    }
+
+    // Implements string.replace(regex, replacement) where the replacement is
+    // a string with $-sequences.
+    pub fn replace_string(
         &mut self,
         activation: &mut Activation<'_, 'gc, '_>,
         text: AvmString<'gc>,
         replacement: AvmString<'gc>,
     ) -> AvmString<'gc> {
+        self.replace_with_fn(activation, &text, |_activation, txt, m| {
+            Self::effective_replacement(&replacement, txt, m)
+        })
+    }
+
+    // Helper for replace_string and replace_function.
+    //
+    // Replaces occurrences of regex with results of f(activation, &text, &match)
+    fn replace_with_fn<F>(
+        &mut self,
+        activation: &mut Activation<'_, 'gc, '_>,
+        text: &AvmString<'gc>,
```

Comment:
For simplicity, this can be changed to `text: &WStr`. Then `effective_replacement` will have to be changed accordingly.

Reply:
But find_utf16_match() expects AvmString, so we'd have to construct one?

Reply:
Right, so let's keep this unchanged.